### ISSUE TITLE
Throwback to #169 :: strange xinput behavior, if any

### ISSUE THREAD

**Issue by volumetricsteve** [yquake2/yquake2#353]
Hello again,

I've once more run into the goofy behavior I found in issue #169 but this time on a gaming desktop pc with both usb and ps/2 keyboards.  So previously on yquake2, we uncovered that it was basically some weird thing that icewm was doing that prevented me from rocket jumping (and probably other fancy maneuvers as well)  

I saw one of the tricks I did back then to test without icewm was to do::

./startx quake2

Curiously, I find that on the laptop for which I placed #169, I have yearly snapshots of Yquake2.   The 2016 snapshot, which is version 6.01pre allows me to run this way.  Both 7.10 and 7.30 both open this way, but they don't work.  In both of the latter cases, the keyboard - any keyboard, seems to be dead...or null....the system seems totally unaware of any keyboard key presses.  The mouse, in all cases, however, works perfectly.

This is a tall order, but what may have changed between 6.01 pre and 7.10&7.30 that is stopping my keyboards from working?

I'm still on Crux, IceWM is still my window manager, I've forgotten how to check for my version of SDL and the like... 

Happy New Year!

**Comment by DanielGibson** [comment]
I don't think we made any keyboard input related changes.
You should check the SDL2 version on your systems. In case you built SDL2 yourself (or it was compiled on your system by the distro's package manager) make sure that all of SDL2 built dependencies are available for building

**Comment by volumetricsteve** [comment]
I'm sorry for the late reply.  I finally got around to really fleshing this out and I've noticed something very strange.  My yquake2 that "just works" was compiled -completely- differently from any other version I have.

when I run ldd against that binary (version 6.01pre), I see:
        linux-vdso.so.1 (0x00007fff36798000)
        libm.so.6 => /lib/libm.so.6 (0x00007f13d2221000)
        libdl.so.2 => /lib/libdl.so.2 (0x00007f13d201d000)
        libvorbis.so.0 => /usr/lib/libvorbis.so.0 (0x00007f13d1df2000)
        libvorbisfile.so.3 => /usr/lib/libvorbisfile.so.3 (0x00007f13d1bea000)
        libogg.so.0 => /usr/lib/libogg.so.0 (0x00007f13d19e4000)
        libz.so.1 => /lib/libz.so.1 (0x00007f13d17cd000)
        libGL.so.1 => /usr/lib/libGL.so.1 (0x00007f13d1481000)
        libSDL2-2.0.so.0 => /lib/libSDL2-2.0.so.0 (0x00007f13d1113000)
        libc.so.6 => /lib/libc.so.6 (0x00007f13d0d75000)
        /lib/ld-linux-x86-64.so.2 (0x00007f13d2525000)
        libnvidia-tls.so.340.102 => /usr/lib/tls/libnvidia-tls.so.340.102 (0x00007f13d0b72000)
        libnvidia-glcore.so.340.102 => /usr/lib/libnvidia-glcore.so.340.102 (0x00007f13cdf5e000)
        libX11.so.6 => /usr/lib/libX11.so.6 (0x00007f13cdc23000)
        libXext.so.6 => /usr/lib/libXext.so.6 (0x00007f13cda11000)
        libpthread.so.0 => /lib/libpthread.so.0 (0x00007f13cd7f4000)
        librt.so.1 => /lib/librt.so.1 (0x00007f13cd5ec000)
        libxcb.so.1 => /usr/lib/libxcb.so.1 (0x00007f13cd3c7000)
        libXau.so.6 => /usr/lib/libXau.so.6 (0x00007f13cd1c3000)
        libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x00007f13ccfbd000)

When I run ldd against -any- other yquake2, (the latest one released on feb 7th this month):
        linux-vdso.so.1 (0x00007fffd9d6b000)
        libm.so.6 => /lib/libm.so.6 (0x00007f5b3735f000)
        libdl.so.2 => /lib/libdl.so.2 (0x00007f5b3715b000)
        libSDL2-2.0.so.0 => /aux2/lib/libSDL2-2.0.so.0 (0x00007f5b36ded000)
        libc.so.6 => /lib/libc.so.6 (0x00007f5b36a4f000)
        /lib/ld-linux-x86-64.so.2 (0x00007f5b37663000)
        libpthread.so.0 => /lib/libpthread.so.0 (0x00007f5b36832000)
        librt.so.1 => /lib/librt.so.1 (0x00007f5b3662a000)


I find it very odd that the one that works correctly is linked against a lot more libraries, I imagine the X11 libraries are probably my magic bullet.

However, I'm baffled as to how I ever built 6.01pre with X libs because there's no configure script and I don't see any way to "enable" them...yet somehow...there they are.

For this experiment, I went and got the latest Yq2 and latest SDL (at this time 2.0.9)

I compiled SDL with X (--with-x)
then compiled yq2 linking against that version of SDL (I exported CPATH, LIBRARY_PATH, and LD_LIBRARY_PATH prior to compilation, hopefully that's enough?)

So basically, I think it boils down to being able to link against X libraries, but I have no idea how I managed to do it in 6.01pre without a configure script..and that was back in 2016 so I still barely had any idea what I was doing.  

Does any of this make any sense?  How could I possibly have linked against the X libs?

**Comment by DanielGibson** [comment]
the X libs come from SDL2, which should be linked against them.
A difference I see is `/aux2/lib/libSDL2-2.0.so.0` vs `/lib/libSDL2-2.0.so.0`.
When `./configure`-ing SDL2 it should show which features and libs it'll use.
On my system it looks like
```
SDL2 Configure Summary:
Building Shared Libraries
Building Static Libraries
Enabled modules : atomic audio video render events joystick haptic sensor power filesystem threads timers file loadso cpuinfo assembly
Assembly Math   : mmx 3dnow sse sse2 sse3
Audio drivers   : disk dummy oss alsa(dynamic) pulse(dynamic) sndio
Video drivers   : dummy x11(dynamic) opengl opengl_es2 vulkan wayland(dynamic)
X11 libraries   : xcursor xdbe xinerama xinput2 xinput2_multitouch xrandr xscrnsaver xshape xvidmode
Input drivers   : linuxev linuxkd
Using libsamplerate : NO
Using libudev       : YES
Using dbus          : YES
Using ime           : YES
Using ibus          : YES
Using fcitx         : NO
```
You probably don't need all of those libs, but you should make sure to at least have the X11 libs and input drivers (if you don't, some dependency is missing).

Note that by default SDL2 (at least newer versions of it) does *not* explicitly *link against* most of those libs, so ldd won't show them (it uses `dlopen()` and `dlsym()` to load+use them).
It still needs the development packages of them installed for building though as it uses the their headers.

**Comment by volumetricsteve** [comment]
I wonder if my distribution has let me down and ...I've probably updated X on that laptop just the once and maybe they removed some of the things I've unknowingly been needing here.  To give a more complete answer I'll have to get back into my laptop tomorrow but at a glance I think your SDL 'sees' more X11 things to link against than mine does.  Thanks for still kicking this around, I know it's a crazy rabbit hole to keep going down.

**Comment by volumetricsteve** [comment]
I'm baffled.  My SDL2 build sees the -exact- same X libs yours does.

X11 libraries   : xcursor xdbe xinerama xinput2 xinput2_multitouch xrandr xscrnsaver xshape xvidmode
Input drivers   : linuxev linuxkd

To recap, you're able to do the "startx ./quake2" stunt with your binaries, yeah?

If so, that means there's something more wrong with my build environment than I realized.

I'd also be curious to see what version of Xorg you're running (if it works)  and...what nvidia driver revision you're on..

That's the other weird thing regarding ldd output from my 6.01pre binary..note the presense of the nvidia objects:
        libnvidia-tls.so.340.102 => /usr/lib/tls/libnvidia-tls.so.340.102 (0x00007f77b2ddf000)
        libnvidia-glcore.so.340.102 => /usr/lib/libnvidia-glcore.so.340.102 (0x00007f77b01cb000)

I don't even know where to start with that or how I managed to link against it.  

There has never been a ./configure script with Yamagi Quake2, right?  I'm just at a complete loss with how I got the quake2 binary to link against things..that there seem to have been no way to do....

I mean your code is written for SDL....you're not reaching out to Nvidia, nor are you reaching out to X...then SDL reaches out to X....I don't think I've ever seen it link against anything from nvidia like that...yet here I have /your/ binary that wants both....and works...and runs perfectly.   

Do you still have the source for 6.01pre bouncing around somewhere?  

I figured 6 would be close enough (which I found in your archives) but it can't build and hits a bunch of errors that are unclear..and I feel like that's reaching anyway...there should be some way to get this to work in the current versions if you haven't changed the inputs code much since then.

I have to seriously wonder if I found someone's hacked/mangled version of SDL2 and unknowingly used that..and in some Freakazoidian accident created the perfect quake2 executable.

**Comment by DanielGibson** [comment]
nvidia 396.54
Xorg 1.19.6

There has never been a configure scripft for YQ2, and no, linking it against additional X libs really shouldn't make a difference as SDL should handle that.
There is a compile flag for SDL2 that disables the aforementioned dynamic loading of dependencies (with dlopen()/dlsym()) and links against stuff instead, you could try that to be extra sure.

https://github.com/yquake2/yquake2/releases has the sources of old releases

Maybe it's not YQ2's or SDL's fault after all and something is broken on your system.
TBH, just use another distro, maybe some Ubuntu flavor or Debian.

**Comment by volumetricsteve** [comment]
I have more data points to work with now.

Some applications work fine with the "startx ./whatever" 

gzdoom loads video but the menu engine breaks and it gets stuck on "pause" mouse input works, keyboard does not

rbdoom3 loads, gui menus work, mouse works but again no keyboard 

darkplaces-sdl works flawlessly
quakespasm works flawlessly

All of these were compiled from source on the same machine.

I can't say for absolute certainty what xorg they may have been built with or what version of gcc I built them with...but i'm not entirely convinced that should matter.  

I managed to build yq2/sdl2 late last night, as you suggested without the dlopen...option that I admittedly don't entirely understand.  The funny thing is that makes the ldd output of quake2 look -much- more like my magic 2016 build, but since i guess it has some baring on dynamically loaded objects, sdl fails to load libGL.so for all the obvious reasons.

I'm trying to go the route of static linking to avoid that, or to see if I can throw -lGL on the linker line when quake2 gets built.


i'm also going to see if I can rebuild quakespasm or darkplaces again to see if I can make the magic happen again...it could be that my build environment was just better once and I screwed it up somehow (though I have no clue what would have changed)  and if I can get those working again...there must be some fundamental difference between how they talk to X and how quake2 is talking to X

if not...I guess I'll throw a new hard drive in that machine and go from scratch to see what else might be going on.



yq2, we know the story there

ioquake3 loads but no mouse or keyboard

**Comment by Yamagi** [comment]
There's no need link anything static. The `quake2` binary has an RPATH to the `lib/` directory next to it. Just create a new dir `lib/` next to the binary, copy you libs in and the dynamic linker will prefer them to the system wide libs.

Nevertheless, this look like a general problem affecting most SDL games on Crux. So the Crux support boards, lists, IRC channels, etc. may a more appropriate place for this problem.

**Comment by volumetricsteve** [comment]
Hey Yamagi,

I think you may be right.  I was actually about to go talk to them, so good call.

It's just so weird that it's clearly some SDL games...and clearly some versions.

I'm sure the Crux folks will have something to add.

**Comment by volumetricsteve** [comment]
I've worked it out and I'm all but convinced it has to do with the order in which things are linked at build time.

I've confirmed SDL (Crux's standard issue SDL) is not the issue.  I went and got the quakespasm source code, rebuilt it, and it still works like a charm.  No missed mouse events, no problems rocket jumping, keyboard works perfectly.

I've built it with both SDL1 and 2(standard Crux, I've trashed my own at this point), same successful result each time.

I highly suspect since I'm probably one of a handful of people on earth who care about being able to run things "startx ./quake2" style, all those other engines/games/applications either work or don't purely by accident.

I just now got your quake2 client to correctly initialize video in openGL & grab mouse input....just not keyboard input.

I modified the Makefile a bit but I think the magic was really in this section:

release/quake2 : $(CLIENT_OBJS)
        @echo "===> LD $@"
        ${Q}$(CC) $(CLIENT_OBJS) -lm -lGL $(LDFLAGS) $(SDLLDFLAGS) -o $@
endif

Where I just tossed in -lm -lGL before any other outside objects.  I saw that same order in quakespasm's makefile, so that seems to work on the libGL.so side of things.

The only issue now is figuring out what's going on with the keyboard.  For that I may need to hit an SDL forum, but I thought you would be interested to know I'm making some progress here.

**Comment by volumetricsteve** [comment]
It's not often I get to ask the author something like this, but I wonder if you would be ok with me forking your client and repurposing for my own needs.  Then I'd promise to only bug you with issues -not- pertaining to strange input behavior:)

**Comment by Yamagi** [comment]
I have no problem with someone forking the code. As long as the license (standard GPL2 for most parts, see LICENSE for details) is honored, of course. But you should be aware that Quake II is a mess, even after years of cleanup and thousands of bugfixes the codebase is still very fragile. The biggest problem is that the game data was build against the broken, hacky code, most well-meant changes introduce subtile problems.
